File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        idxToRemove = set()
        stack = []
        for i, c in enumerate(s):
            if c not in '()':
                continue
            elif c == '(':
                stack.append(i)
            elif not stack:
                idxToRemove.add(i)
            else:
                stack.pop()
        
        idxToRemove = idxToRemove.union(set(stack))
        stringBuilder = []
        
        for i, c in enumerate(s):
            if i not in idxToRemove:
                stringBuilder.append(c)
        
        return ''.join(stringBuilder)

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        Set<Integer> idxToRemove = new HashSet<>();
        Deque<Integer> stack = new ArrayDeque<>();
        
        for(int i=0; i<s.length(); i++){
            if(s.charAt(i)=='('){
                stack.push(i);
            } if(s.charAt(i)==')'){
                if(stack.isEmpty()){
                    idxToRemove.add(i);
                } else{
                    stack.pop();
                }
            } 
        }
        
        while(!stack.isEmpty()){
            idxToRemove.add(stack.pop());
        }
        StringBuilder sb = new StringBuilder();
        for(int i=0; i<s.length();i++) {
            if(!idxToRemove.contains(i)){
                sb.append(s.charAt(i));
            }
            
        }
        return sb.toString();

    }
}